src/modules/manual_control/ManualControlSelectorTest.cpp

```diff
@@ -0,0 +1,198 @@
+	selector.update_manual_control_input(timestamp, input, 1);
+
+	EXPECT_TRUE(selector.setpoint().valid);
+	EXPECT_TRUE(selector.setpoint().chosen_input.data_source == manual_control_input_s::SOURCE_MAVLINK_3);
+	EXPECT_EQ(selector.instance(), 1);
+
+	// But only the first MAVLink source wins, others are too late.
+
+	timestamp += 100_ms;
+
+	// Now provide input with the correct source.
+	input.data_source = manual_control_input_s::SOURCE_MAVLINK_4;
+	input.timestamp_sample = timestamp;
+	selector.update_manual_control_input(timestamp, input, 1);
+
+	EXPECT_TRUE(selector.setpoint().valid);
+	EXPECT_TRUE(selector.setpoint().chosen_input.data_source == manual_control_input_s::SOURCE_MAVLINK_3);
+	EXPECT_EQ(selector.instance(), 1);
+}
+
+TEST(ManualControlSelector, AutoInput)
+{
+	ManualControlSelector selector;
+	selector.set_rc_in_mode(3);
+	selector.set_timeout(500_ms);
+
+	uint64_t timestamp = some_time;
+
+	manual_control_input_s input {};
+	input.data_source = manual_control_input_s::SOURCE_RC;
+	input.timestamp_sample = timestamp;
+	selector.update_manual_control_input(timestamp, input, 0);
+
+	EXPECT_TRUE(selector.setpoint().valid);
+	EXPECT_TRUE(selector.setpoint().chosen_input.data_source == manual_control_input_s::SOURCE_RC);
+	EXPECT_EQ(selector.instance(), 0);
+
+	timestamp += 100_ms;
+
+	// Now provide input from MAVLink as well which should get ignored.
+	input.data_source = manual_control_input_s::SOURCE_MAVLINK_0;
+	input.timestamp_sample = timestamp;
+	selector.update_manual_control_input(timestamp, input, 1);
+
+	EXPECT_TRUE(selector.setpoint().valid);
+	EXPECT_TRUE(selector.setpoint().chosen_input.data_source == manual_control_input_s::SOURCE_RC);
+	EXPECT_EQ(selector.instance(), 0);
+
+	timestamp += 500_ms;
+
+	// Now we'll let RC time out, so it should switch to MAVLINK.
+	input.data_source = manual_control_input_s::SOURCE_MAVLINK_0;
+	input.timestamp_sample = timestamp;
+	selector.update_manual_control_input(timestamp, input, 1);
+
+	EXPECT_TRUE(selector.setpoint().valid);
+	EXPECT_TRUE(selector.setpoint().chosen_input.data_source == manual_control_input_s::SOURCE_MAVLINK_0);
+	EXPECT_EQ(selector.instance(), 1);
+}
+
+TEST(ManualControlSelector, RcTimeout)
+{
+	ManualControlSelector selector;
+	selector.set_rc_in_mode(0);
+	selector.set_timeout(500_ms);
+
+	uint64_t timestamp = some_time;
+
+	manual_control_input_s input {};
+	input.data_source = manual_control_input_s::SOURCE_RC;
+	input.timestamp_sample = timestamp;
+	selector.update_manual_control_input(timestamp, input, 0);
+
+	EXPECT_TRUE(selector.setpoint().valid);
+	EXPECT_TRUE(selector.setpoint().chosen_input.data_source == manual_control_input_s::SOURCE_RC);
+	EXPECT_EQ(selector.instance(), 0);
+
+	timestamp += 600_ms;
+
+	// Update, to make sure it notices the timeout
+	selector.update_time_only(timestamp);
+
+	EXPECT_FALSE(selector.setpoint().valid);
+	EXPECT_EQ(selector.instance(), -1);
+}
+
+TEST(ManualControlSelector, RcOutdated)
```

Comment:
I added this test to show that you could end up accepting outdated data if it arrives already being older than the timeout. That's happening because checking the timeout only happens for the signal that's already in use so you end up invalidating the old outdated one and accepting the new outdated one in every iteration.

Reply:
Good catch. Fixed it and test is now passing :tada:.